include/nil/crypto3/merkle/proof.hpp

```diff
@@ -77,28 +85,25 @@ namespace nil {
 
                     for (size_t cur_row = 0; cur_row < path.size(); ++cur_row) {
 
-                        std::array<uint8_t, (value_bits / std::numeric_limits<std::uint8_t>::digits +
-                                             (value_bits % std::numeric_limits<std::uint8_t>::digits ? 1 : 0)) *
-                                                arity>
-                            new_input;
+                        nil::crypto3::accumulator_set<hash_type> acc;
                         size_t missing_idx = arity - 1;    // If every previous index was fine - missing the last one.
+                        std::array<std::size_t, arity> position_map;
 
                         for (size_t i = 0; i < arity - 1; ++i) {
-                            std::copy(path[cur_row][i].hash.begin(), path[cur_row][i].hash.end(),
-                                      new_input.begin() +
-                                          path[cur_row][i].position *
-                                              (value_bits / std::numeric_limits<std::uint8_t>::digits +
-                                               (value_bits % std::numeric_limits<std::uint8_t>::digits ? 1 : 0)));
+                            position_map[path[cur_row][i].position] = i;
```

Comment:
It is proposed to avoid an additional loop and complex calls to array elements by sacrificing a little code duplication.
We can use something like this
```
for (size_t i = 0; i < arity - 1; ++i) {
  if (path[cur_row][i].position != i && !was_missing) {
    crypto3::hash<hash_type>(d.begin(), d.end(), acc);
    was_missing = true;
  } 
  crypto3::hash<hash_type>(path[cur_row][i].hash.begin(), path[cur_row][i].hash.end(), acc);
}
if (!was_missing) {
  crypto3::hash<hash_type>(d.begin(), d.end(), acc);
}
```

Reply:
Can't select a line 63/71 so I will write here.
It is proposed to solve the problem of determining in what order the positions are on the same level of the tree at the moment of creating the authentication path.
```
if (cur_leaf != current_child) {
  std::size_t save_position = current_child % arity;
  if (save_position > cur_leaf_pos) {
    --save_position;
  }
  path[cur_row][save_position] = path_element_type(tree[current_child], current_child % arity);
}
```